[Optimization] Add Fused Triton Kernel for GPT-OSS Router

ijpq opened the pull request:
<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
Collecting environment information...
==============================
        System Info
==============================
OS                           : Ubuntu 24.04.3 LTS (x86_64)
GCC version                  : (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version                : Could not collect
CMake version                : version 3.28.3
Libc version                 : glibc-2.39

==============================
       PyTorch Info
==============================
PyTorch version              : 2.9.0+cu128
Is debug build               : False
CUDA used to build PyTorch   : 12.8
ROCM used to build PyTorch   : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.11 | packaged by conda-forge | (main, Jun  4 2025, 14:45:31) [GCC 13.3.0] (64-bit runtime)
Python platform              : Linux-5.4.0-216-generic-x86_64-with-glibc2.39

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : 12.8.93
CUDA_MODULE_LOADING set to   : 
GPU models and configuration : GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version        : 570.153.02
cuDNN version                : Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.10.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : True

==============================
          CPU Info
==============================
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             64
On-line CPU(s) list:                0-63
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 7343 16-Core Processor
CPU family:                         25
Model:                              1
Thread(s) per core:                 2
Core(s) per socket:                 16
Socket(s):                          2
Stepping:                           1
Frequency boost:                    enabled
CPU(s) scaling MHz:                 86%
CPU max MHz:                        3200.0000
CPU min MHz:                        1500.0000
BogoMIPS:                           6387.85
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca
Virtualization:                     AMD-V
L1d cache:                          1 MiB (32 instances)
L1i cache:                          1 MiB (32 instances)
L2 cache:                           16 MiB (32 instances)
L3 cache:                           256 MiB (8 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-15,32-47
NUMA node1 CPU(s):                  16-31,48-63
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.5.2
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cudnn-frontend==1.16.0
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-cufile-cu12==1.13.1.3
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-cutlass-dsl==4.3.0
[pip3] nvidia-ml-py==13.580.82
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvshmem-cu12==3.3.20
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] pyzmq==27.1.0
[pip3] torch==2.9.0
[pip3] torchaudio==2.9.0
[pip3] torchvision==0.24.0
[pip3] transformers==4.57.1
[pip3] triton==3.5.0
[conda] flashinfer-python         0.5.2                    pypi_0    pypi
[conda] numpy                     2.2.6                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.8.4.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.8.90                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.8.93                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.8.90                  pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.10.2.21                pypi_0    pypi
[conda] nvidia-cudnn-frontend     1.16.0                   pypi_0    pypi
[conda] nvidia-cufft-cu12         11.3.3.83                pypi_0    pypi
[conda] nvidia-cufile-cu12        1.13.1.3                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.9.90                pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.7.3.90                pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.8.93                pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.7.1                    pypi_0    pypi
[conda] nvidia-cutlass-dsl        4.3.0                    pypi_0    pypi
[conda] nvidia-ml-py              13.580.82                pypi_0    pypi
[conda] nvidia-nccl-cu12          2.27.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.8.93                  pypi_0    pypi
[conda] nvidia-nvshmem-cu12       3.3.20                   pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.8.90                  pypi_0    pypi
[conda] pyzmq                     27.1.0                   pypi_0    pypi
[conda] torch                     2.9.0                    pypi_0    pypi
[conda] torchaudio                2.9.0                    pypi_0    pypi
[conda] torchvision               0.24.0                   pypi_0    pypi
[conda] transformers              4.57.1                   pypi_0    pypi
[conda] triton                    3.5.0                    pypi_0    pypi

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.1.dev11359+gf3c61a92a.d20251124 (git sha: f3c61a92a, date: 20251124)
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled
GPU Topology:
  	^[[4mGPU0	NIC0	CPU Affinity	NUMA Affinity	GPU NUMA ID^[[0m
GPU0	 X 	SYS	16-31,48-63	1		N/A
NIC0	SYS	 X 				

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx4_0

==============================
     Environment Variables
==============================
NVIDIA_VISIBLE_DEVICES=D.cca8b812a3ffc614142fc7b8d7c8b2bc17a7a24d02ca5110609b5ae90135abe8/gpu=7
NVIDIA_REQUIRE_CUDA=cuda>=12.8 brand=unknown,driver>=470,driver<471 brand=grid,driver>=470,driver<471 brand=tesla,driver>=470,driver<471 brand=nvidia,driver>=470,driver<471 brand=quadro,driver>=470,driver<471 brand=quadrortx,driver>=470,driver<471 brand=nvidiartx,driver>=470,driver<471 brand=vapps,driver>=470,driver<471 brand=vpc,driver>=470,driver<471 brand=vcs,driver>=470,driver<471 brand=vws,driver>=470,driver<471 brand=cloudgaming,driver>=470,driver<471 brand=unknown,driver>=535,driver<536 brand=grid,driver>=535,driver<536 brand=tesla,driver>=535,driver<536 brand=nvidia,driver>=535,driver<536 brand=quadro,driver>=535,driver<536 brand=quadrortx,driver>=535,driver<536 brand=nvidiartx,driver>=535,driver<536 brand=vapps,driver>=535,driver<536 brand=vpc,driver>=535,driver<536 brand=vcs,driver>=535,driver<536 brand=vws,driver>=535,driver<536 brand=cloudgaming,driver>=535,driver<536 brand=unknown,driver>=550,driver<551 brand=grid,driver>=550,driver<551 brand=tesla,driver>=550,driver<551 brand=nvidia,driver>=550,driver<551 brand=quadro,driver>=550,driver<551 brand=quadrortx,driver>=550,driver<551 brand=nvidiartx,driver>=550,driver<551 brand=vapps,driver>=550,driver<551 brand=vpc,driver>=550,driver<551 brand=vcs,driver>=550,driver<551 brand=vws,driver>=550,driver<551 brand=cloudgaming,driver>=550,driver<551 brand=unknown,driver>=560,driver<561 brand=grid,driver>=560,driver<561 brand=tesla,driver>=560,driver<561 brand=nvidia,driver>=560,driver<561 brand=quadro,driver>=560,driver<561 brand=quadrortx,driver>=560,driver<561 brand=nvidiartx,driver>=560,driver<561 brand=vapps,driver>=560,driver<561 brand=vpc,driver>=560,driver<561 brand=vcs,driver>=560,driver<561 brand=vws,driver>=560,driver<561 brand=cloudgaming,driver>=560,driver<561 brand=unknown,driver>=565,driver<566 brand=grid,driver>=565,driver<566 brand=tesla,driver>=565,driver<566 brand=nvidia,driver>=565,driver<566 brand=quadro,driver>=565,driver<566 brand=quadrortx,driver>=565,driver<566 brand=nvidiartx,driver>=565,driver<566 brand=vapps,driver>=565,driver<566 brand=vpc,driver>=565,driver<566 brand=vcs,driver>=565,driver<566 brand=vws,driver>=565,driver<566 brand=cloudgaming,driver>=565,driver<566
NCCL_VERSION=2.25.1-1
NVIDIA_DRIVER_CAPABILITIES=all
NVIDIA_PRODUCT_NAME=CUDA
CUDA_VERSION=12.8.1
LD_LIBRARY_PATH=/usr/local/cuda/lib64
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1


```

</details>

## Purpose

Resolves: #28986

1. add a fused topk+softmax triton kernel for gptoss and others.
2. minimize modification in model's interface, by indicating custom_routing_function.


## Test Plan

compare results with torch and intact impl.
- [x]  It's necessary to roofline. to prove performance is better than existed cuda kernel, including num experts is power of 2 or not. 
## Test Result


---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

ijpq committed:
[Optimization] Add Fused Triton Kernel for topk+softmax

Signed-off-by: ijpq <509634578tk@gmail.com>

ZJY0516 reviewed (commented):
Could you please add a test for this?

ijpq commented on vllm/model_executor/models/gpt_oss.py:
```diff
@@ -185,9 +185,18 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+            )
+
+            x = self.experts(
+                hidden_states=x, topk_weights=topk_weights, topk_ids=topk_indices
```
Sorry, I didn't notice this here. I'll fix it.

mgoin commented on vllm/model_executor/models/gpt_oss.py:
```diff
@@ -174,6 +176,11 @@ def __init__(
             is_sequence_parallel=self.is_sequence_parallel,
+            custom_routing_function=(
+                gpt_oss_custom_routing_function
+                if not current_platform.is_rocm()
```
nit: just check cuda to help other platforms
```suggestion
                if current_platform.is_cuda()
```

mgoin commented on tests/kernels/moe/test_gpt_oss_routing_consistency.py:
nit: could you consolidate these tests into one file?

ijpq replied:
ack

ijpq commented on vllm/model_executor/layers/fused_moe/gpt_oss_fused_router.py:
```diff
@@ -0,0 +1,112 @@
+
+    topk_padded = triton.next_power_of_2(topk)
+
+    grid = (M,)
```
In fact, I don't quite understand the purpose of replacing a handwritten CUDA kernel with Triton in models where the number of experts is a power of 2. If the CUDA kernel is already highly optimized, it should theoretically be difficult for Triton to outperform it in terms of performance. In other words, we now hope to fuse GPTOSS's topksoftmax together, but as far as I can see, the current implementation is already a single kernel. BTW, I just posted perf collected from nsys.
Either, purpose is to just replace it since triton is easier to maintain, in the future.

ijpq commented on tests/kernels/moe/test_gpt_oss_fused_router.py:
```diff
@@ -0,0 +1,29 @@
+@pytest.mark.parametrize("M", [1, 32, 128, 2048])
+@pytest.mark.parametrize("N", [32, 65, 128])
+@pytest.mark.parametrize("topk", [1, 2, 3, 4, 5])
+def test_fused_router(M, N, topk):
```
ack

ijpq commented on tests/kernels/moe/test_gpt_oss_routing_consistency.py:
```diff
@@ -0,0 +1,61 @@
+@pytest.mark.parametrize("num_tokens", [10, 128, 1024])
+@pytest.mark.parametrize("num_experts", [32, 65, 128])
+@pytest.mark.parametrize("topk", [1, 2, 3, 4, 5])
+def test_routing_consistency(num_tokens, num_experts, topk):
```
ack

ijpq committed:
[Optimization]: Optimize Fused Triton Kernel for topk+softmax

- split two kernels, in case renorm or not
- add online softmax
- unroll along M

Signed-off-by: ijpq <509634578tk@gmail.com>

ijpq commented:
**TL;DR**
Achieved point moved to the upper right.

Achieved better flop/s and get rid of mem bound in roofline analysis, in commit : 5af77ad
It seems error in topkGatingSoftmax(original impl) when renorm close, because it failed to match the result from torch impl. Requires investigation further.
<img width="1715" height="392" alt="image" src="https://github.com/user-attachments/assets/7af3f6cf-3088-4774-9753-a2619c35b3b3" />

these two roofline I collected using 
```
ncu --set full --target-processes all -o triton_topk_profile   --launch-skip 100 --launch-count 5   --kernel-name _topk_softmax_kernel vllm bench latency     --model openai/gpt-oss-20b     --num-iters-warmup 5     --num-iters 1     --batch-size 16     --input-len 1024     --output-len 256
```

ijpq committed:
[Fix]: Tweak a few little things in triton kernel

- delete unnecessary splited kernel
- add skipif in unittest
- ACK reviews

Signed-off-by: ijpq <509634578tk@gmail.com>

ijpq committed:
[Optimization]: add specialization for small topk

Signed-off-by: ijpq <509634578tk@gmail.com>

ijpq committed:
[Optimization]: unroll for each program along M

Signed-off-by: ijpq <509634578tk@gmail.com>

ijpq committed:
[Optimization]: add autotune and fix little things

Signed-off-by: ijpq <509634578tk@gmail.com>

ijpq commented:
`vllm bench latency     --model openai/gpt-oss-20b     --num-iters-warmup 20     --num-iters 5     --batch-size 16     --input-len 4096     --output-len 16`

original cuda kernel:
```
Avg latency: 9.52765977279996 seconds
10% percentile latency: 9.498314082200341 seconds
25% percentile latency: 9.513294739999765 seconds
50% percentile latency: 9.520952906000275 seconds
75% percentile latency: 9.555723011999362 seconds
90% percentile latency: 9.558289942199554 seconds
99% percentile latency: 9.559830100319669 seconds
```

triton:
```
Avg latency: 9.327314970800217 seconds
10% percentile latency: 9.278146413800096 seconds
25% percentile latency: 9.291758477000258 seconds
50% percentile latency: 9.34792246699999 seconds
75% percentile latency: 9.360818809000193 seconds
90% percentile latency: 9.36452956120047 seconds
99% percentile latency: 9.366756012520636 seconds

```

ijpq committed:
[Optimization]: further specialize for M,N,topk

Signed-off-by: ijpq <509634578tk@gmail.com>

ijpq committed:
[Optimization]: add bitonic sort within warp

Signed-off-by: ijpq <509634578tk@gmail.com>

shaginhekvs commented on vllm/model_executor/layers/fused_moe/triton_bitonic_sort.py:
```diff
@@ -0,0 +1,146 @@
+def bitonic_ce_descending(
+    val, idx, stride: tl.constexpr, log2_length_pair: tl.constexpr
+):
+    new_val, new_idx = tl.inline_asm_elementwise(
```
it would be good to describe how this PTX asm was generated, and maybe the .cu code in test file to help reproduce it....

bitonic_ce_descending description would also make things clearer.

ijpq replied:
The reason for using inline asm is I find the previous iterative top-k algorithm unclear in terms of what the Triton compiler specifically does. However, it's evident that when sorting elements within a warp, we can utilize `shfl_sync` to bypass shared memory. At the Triton kernel level, I'm only able to write PTX instead of calling `shfl_sync` like in CUDA C.

ijpq replied:
I agree with adding clearer comments about whole bitnoic sort code. Besides, what specific .cu code you referred to?

ijpq commented on vllm/model_executor/layers/fused_moe/gpt_oss_fused_router.py:
```diff
@@ -50,6 +39,7 @@ def _topk_softmax_kernel(
     RENORM: tl.constexpr,
     ROWS_PER_PID: tl.constexpr,
     num_stages: tl.constexpr,
+    USE_BITONIC: tl.constexpr,
```
I referred to https://www.geeksforgeeks.org/dsa/bitonic-sort/, and it seems that the implementation in the [anilshanbhag/gpu-topk](https://github.com/anilshanbhag/gpu-topk) is also based on this, since this repo isn't the first place where it'd appeared, according to https://en.wikipedia.org/wiki/Bitonic_sorter. I suppose they implemented topk by using bitonic sort algo across warp even threadblock. Anyway, I feel that it's unnecessary to cite this repository in this case？
I'll add some comments to this algo and inline asm.

shaginhekvs replied:
yes I agree, geeksforgeeks should be enough, explanation would be helpful thanks

ijpq committed:
[RFC]: add comments to explain descending bitonic sort in inline PTX

Signed-off-by: ijpq <509634578tk@gmail.com>
